parser-typechecker/src/Unison/Codebase/FileCodebase/SlimCopyRegenerateIndex.hs

```diff
@@ -0,0 +1,305 @@
+import           Control.Monad.Writer           ( MonadWriter, execWriterT )
+import qualified Control.Monad.Writer          as Writer
+import           Control.Monad.Except           ( MonadError, runExceptT, throwError )
+
+import           UnliftIO.Directory             ( doesFileExist )
+import           Unison.Codebase                ( CodebasePath )
+import qualified Unison.Codebase.Causal        as Causal
+import           Unison.Codebase.Branch         ( Branch(..) )
+import qualified Unison.Codebase.Branch        as Branch
+import qualified Unison.Codebase.Branch.Dependencies as BD
+import qualified Unison.Codebase.Patch         as Patch
+import qualified Unison.Codebase.Serialization as S
+import qualified Unison.Codebase.Serialization.V1 as V1
+import qualified Unison.Codebase.TermEdit      as TermEdit
+import qualified Unison.Codebase.TypeEdit      as TypeEdit
+import qualified Unison.DataDeclaration        as DD
+import qualified Unison.LabeledDependency      as LD
+import           Unison.Reference               ( Reference )
+import qualified Unison.Reference              as Reference
+import qualified Unison.Referent               as Referent
+import qualified Unison.Term                   as Term
+import           Unison.Type                    ( Type )
+import qualified Unison.Type                   as Type
+import           Unison.Var                     ( Var )
+import qualified Unison.Util.Relation          as Relation
+import           Unison.Util.Relation           ( Relation )
+import           Unison.Util.Monoid (foldMapM)
+
+import Data.Monoid.Generic
+import Unison.Codebase.FileCodebase.Common
+
+data SyncedEntities = SyncedEntities
+  { _syncedTerms       :: Set Reference.Id
+  , _syncedDecls       :: Set Reference.Id
+  , _syncedEdits       :: Set Branch.EditHash
+  , _syncedBranches    :: Set Branch.Hash
+  , _dependentsIndex   :: Relation Reference Reference.Id
+  , _typeIndex         :: Relation Reference Referent.Id
+  , _typeMentionsIndex :: Relation Reference Referent.Id
+  } deriving Generic
+  deriving Show
+  deriving Semigroup via GenericSemigroup SyncedEntities
+  deriving Monoid via GenericMonoid SyncedEntities
+
+makeLenses ''SyncedEntities
+
+syncToDirectory :: forall m v a
+  . MonadIO m
+  => Var v
+  => S.Format v
+  -> S.Format a
+  -> CodebasePath
+  -> CodebasePath
+  -> Branch m
+  -> m ()
+syncToDirectory fmtV fmtA = syncToDirectory' (S.get fmtV) (S.get fmtA)
+
+data Error
+  = MissingBranch Branch.Hash
+  | MissingPatch Branch.EditHash
+  | MissingTerm Reference.Id
+  | MissingTypeOfTerm Reference.Id
+  | MissingDecl Reference.Id
+  | InvalidBranch Branch.Hash
+  | InvalidTerm Reference.Id
+  | InvalidTypeOfTerm Reference.Id
+  | InvalidDecl Reference.Id
+  deriving Show
+
+syncToDirectory' :: forall m v a
+  . MonadIO m
+  => Var v
+  => S.Get v
+  -> S.Get a
+  -> CodebasePath
+  -> CodebasePath
+  -> Branch m
+  -> m ()
+syncToDirectory' getV getA srcPath destPath newRoot =
```

Comment:
This can be future work, but I am guessing you could get additional speedups by parallelizing / pipelining the work. So have the main thread putting items to write on a queue, and one or more threads pulling from that queue and just copying the files.

Reply:
Good idea.